contracts/tatum/Tatum721Provenance.sol

```diff
@@ -0,0 +1,223 @@
+        internal
+        virtual
+        override(ERC721, ERC721URIStorage)
+    {
+        return ERC721URIStorage._burn(tokenId);
+    }
+
+    function tokenCashbackValues(uint256 tokenId)
+        public
+        view
+        virtual
+        returns (uint256[] memory)
+    {
+        return _cashbackValues[tokenId];
+    }
+
+    function tokenCashbackRecipients(uint256 tokenId)
+        public
+        view
+        virtual
+        returns (address[] memory)
+    {
+        return _cashbackRecipients[tokenId];
+    }
+
+    function updateCashbackForAuthor(uint256 tokenId, uint256 cashbackValue)
+        public
+        returns (bool)
+    {
+        for (uint256 i = 0; i < _cashbackValues[tokenId].length; i++) {
+            if (_cashbackRecipients[tokenId][i] == _msgSender()) {
+                _cashbackValues[tokenId][i] = cashbackValue;
+                return true;
+            }
+        }
+        return true;
+    }
+
+    function burn(uint256 tokenId) public virtual {
+        //solhint-disable-next-line max-line-length
+        require(
+            _isApprovedOrOwner(_msgSender(), tokenId),
+            "ERC721Burnable: caller is not owner nor approved"
+        );
+        _burn(tokenId);
+    }
+
+    function safeTransfer(
+        address to,
+        uint256 tokenId,
+        string memory data,
+        uint256 value
+    ) public payable {
+        bytes memory bytesData = bytes(data);
+        if (_cashbackRecipients[tokenId].length != 0) {
+            // checking cashback addresses exists and sum of cashbacks
+            require(
+                _cashbackRecipients[tokenId].length != 0,
+                "CashbackToken should be of cashback type"
+            );
+            uint256 percentSum = 0;
+            for (uint256 i = 0; i < _cashbackValues[tokenId].length; i++) {
+                percentSum += _cashbackValues[tokenId][i];
+            }
+            uint256 sum = (percentSum * value) / 100;
+            if (sum > msg.value) {
+                payable(msg.sender).transfer(msg.value);
+                revert(
+                    "Value should be greater than or equal to cashback value"
+                );
+            }
+            for (uint256 i = 0; i < _cashbackRecipients[tokenId].length; i++) {
+                // transferring cashback to authors
+                if (_cashbackValues[tokenId][i] > 0) {
+                    payable(_cashbackRecipients[tokenId][i]).transfer(
+                        (_cashbackValues[tokenId][i] * value) / 100
+                    );
+                }
+            }
+            if (msg.value > sum) {
+                payable(msg.sender).transfer(msg.value - sum);
+            }
+            _safeTransfer(_msgSender(), to, tokenId, bytesData);
+        } else {
+            _safeTransfer(_msgSender(), to, tokenId, bytesData);
+            _setTokenData(tokenId, data, value);
+        }
+        emit TransferWithProvenance(tokenId, to, data, value);
```

Comment:
we need to emit 2 events - our Provenance event and Transfer() event from the original contract

Reply:
transfer event is by default